### PR TITLE
FedPAC Baseline

### PR DESCRIPTION
**Pull request by apzl**
## Issue
Implementation of FedPAC as part of SoR
### Related issues/PRs
Issue #2027 

### Checklist

- [ ] Implement proposed change
- [ ] Write tests
- [ ] Update [documentation](https://flower.dev/docs/writing-documentation.html)
- [ ] Update [changelog](https://github.com/adap/flower/blob/main/doc/source/changelog.rst)
- [ ] Make CI checks pass
- [ ] Ping maintainers on [Slack](https://flower.dev/join-slack/) (channel `#contributions`)

### CODE REVIEW

**Comment by jafermarq** [comment]
Hi @apzl, should I do a first pass through your baseline and leave some comments? or do you prefer if I wait one extra day?

**Comment by apzl** [comment]
Hi @jafermarq ,   Please take a look at my baseline. Any feedback from your side would be really helpful. 
I need to do a bit more on the data partition side to replicate the exact experiments from paper. But baseline algorithms are all done.

**Review by jafermarq — COMMENTED** [review]
Hi @apzl,

I just finished doing my initial review of your `FedPac` implementation. I was able to follow your instructions in the readme and run the code. I have left several suggestions to improve the `README.md` and also some small changes to your `pyproject.toml` and `strategy.py`.

In addition to those:
* You can remove the `EXTENDED_README.md` file since its only purpose was to guide you in the initial steps for creating your baseline. 
* Don't forget to run the formatting and testing scripts. The former needs to show no errors before you can run the latter. It's easy! Simply: (1) sourcing your environment from your baseline directory, (2) `cd ..`, (3) `./dev/format-baseline.sh fedpac`. Then do the same with the test script.
* If you include plots, please add them into a directory with name `_static` and place it at the same level as the `README.md`.

Overall, your baseline seems to be going in the right direction! Looking forward to seeing the first results.

Please ping me by writing a comment in this PR once you want me to take another look. If some of the comments below are not clear or you want to discuss them further, happy to chat about it.

File: baselines/fedpac/pyproject.toml [hunk 1]
```diff
@@ -0,0 +1,143 @@
...
+
+[tool.poetry.dependencies]
+python = ">=3.10.0, <3.11.0"
+flwr = { extras = ["simulation"], version = "1.5.0" }
+hydra-core = "1.3.2"
+matplotlib = "3.7.1"
+jupyter = "^1.0.0"
+pandas = "^2.0.3"
+torch = { url = "https://download.pytorch.org/whl/cu117/torch-2.0.1%2Bcu117-cp310-cp310-linux_x86_64.whl"}
+torchvision = { url = "https://download.pytorch.org/whl/cu117/torchvision-0.15.2%2Bcu117-cp310-cp310-linux_x86_64.whl"}
```
Comment (jafermarq) [inline baselines/fedpac/pyproject.toml:47]:
```suggestion
torch = { url = "https://download.pytorch.org/whl/cu116/torch-1.13.1%2Bcu116-cp310-cp310-linux_x86_64.whl"}
torchvision = { url = "https://download.pytorch.org/whl/cu116/torchvision-0.14.1%2Bcu116-cp310-cp310-linux_x86_64.whl"}
```

File: baselines/fedpac/README.md [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+
+
+## About this baseline
+
+****What’s implemented:**** The code in this directory replicates the experiments in *Personalized Federated Learning with Feature Alignment and Classifier Collaboration* (Xu et al., 2023) for EMNIST and CIFAR10, which proposed the FedPAC algorithm. Concretely, it replicates the results for EMNIST and CIFAR10 in Table 1.
+
+****Datasets:**** EMNIST and CIFAR10 from PyTorch's Torchvision
+
+****Hardware Setup:**** These experiments were run on a desktop machine with 24 CPU threads. Any machine with 4 CPU cores or more would be able to run it in a reasonable amount of time. Note: we install PyTorch with GPU support but by default, the entire experiment runs on CPU-only mode.
```
[the remark this answers is not in the document]
Comment (apzl) [reply baselines/fedpac/README.md:25]: On the classifier aggregation step, there are some calculations that I need to test without `cuda`. I'll update the hardware section of readme after some more tests

File: baselines/fedpac/pyproject.toml (hunk 1 above)
Comment (apzl) [reply baselines/fedpac/pyproject.toml:47]: Is torch-1.13 still the default?  Saw torch-2.0 in FedProx.

Comment (jafermarq) [reply baselines/fedpac/pyproject.toml:47]: ok. let's leave it

File: baselines/fedpac/README.md [hunk 3]
```diff
@@ -0,0 +1,99 @@
+---
+title: Personalized Federated Learning with Feature Alignment and Classifier Collaboration
+url: https://openreview.net/forum?id=SXZr8aDKia
+labels: [label1, label2] # please add between 4 and 10 single-word (maybe two-words) labels (e.g. "system heterogeneity", "image classification", "asynchronous", "weight sharing", "cross-silo")
+dataset: [dataset1, dataset2] # list of datasets you include in your baseline
```
Comment (jafermarq) [inline baselines/fedpac/README.md:5]: Don't forget to add meaningful labels for your baseline. Then remove the comments.

File: baselines/fedpac/fedpac/client.py [hunk 4]
```diff
@@ -0,0 +1,240 @@
...
+    def get_feature_extractor(self):
+        """function to extract feature extractor layers"""
+        feature_extractors = {}
+        model=self.net
+        train_data= self.trainloader
+        with torch.no_grad():
+            for inputs, labels in train_data:
+                inputs, labels = inputs.to(self.device), labels.to(self.device)
+                features, outputs = model(inputs)
+                feature_extractor = features.clone().detach()
```
Comment (jafermarq) [inline baselines/fedpac/fedpac/client.py:105]: is a `.cpu()` missing here? I'm getting an error in the strategy when clients train on GPU. (which seem to return centroids that are not on `CPU`).

**Review by jafermarq — COMMENTED** [review]
Hi @apzl,

Thanks for adding those corrections. I just highlighted a couple of small points (see below). I'll do a more in-depth review once the remaining of the code is added. Please ping me

**Comment by WilliamLindskog** [comment]
Hi @apzl, 

Just checking in here. Are you still eager complete this baseline? 

Best regards
William

**Comment by apzl** [comment]
I wish to complete this but I won't be able to complete this immediately.

I can create a new issue later incase you need to close this.

**Comment by WilliamLindskog** [comment]
No need to close the issue then. Thank you for quick reply.

**Comment by WilliamLindskog** [comment]
@apzl Hi again, 

are you still looking to complete this baseline? 

Could you review the changes requested, and push code that is up to date with current codebase? 

Best regards
William
